include/dynd/strided_vals.hpp

```diff
@@ -8,284 +8,195 @@
+                         ->stride);
+    }
+
+    static void incr(const char *arrmeta, const char **data, intptr_t *index)
+    {
+      ++index[0];
+      *data +=
+          reinterpret_cast<const fixed_dim_type_arrmeta *>(arrmeta)->stride;
+    }
+  };
+
+  template <typename T>
+  class fixed_dim<T, false> {
+  public:
+    static const size_t arrmeta_size =
+        sizeof(fixed_dim_type_arrmeta) + T::arrmeta_size;
+    static const intptr_t ndim = 1 + T::ndim;
+    typedef typename T::dtype dtype;
+
+    static int at(const intptr_t *i, const char *arrmeta, const char *data)
+    {
+      return T::at(
+          i + 1, arrmeta + sizeof(size_stride_t),
+          data +
+              i[0] * reinterpret_cast<const size_stride_t *>(arrmeta)->stride);
+    }
+
+    static void incr(const char *arrmeta, const char **data, intptr_t *index)
+    {
+      const size_stride_t *size_stride =
+          reinterpret_cast<const size_stride_t *>(arrmeta);
+
+      if (++index[0] != size_stride->dim_size) {
+        *data += size_stride->stride;
+      } else {
+        index[0] = 0;
+        *data -= (size_stride->dim_size - 1) * size_stride->stride;
+        T::incr(arrmeta + sizeof(size_stride_t), data, index + 1);
+      }
+    }
+  };
+
+} // namespace dynd::detail
 
-    static void incr(const char *&pointer, intptr_t *index, const intptr_t *sizes, const intptr_t *strides) {
-        if (++index[N - 1] != sizes[N - 1]) {
-            pointer += strides[N - 1];
-        } else {
-            index[N - 1] = 0;
-            pointer -= (sizes[N - 1] - 1) * strides[N - 1];
-            strided_utils<N - 1>::incr(pointer, index, sizes, strides);
-        }
-    }
+template <typename T>
+struct is_dim {
+  static const bool value = false;
 };
 
-template <>
-struct strided_utils<1> {
-    static const char *get(const char *pointer, const intptr_t *index, const intptr_t *strides) {
-        return pointer + index[0] * strides[0];
-    }
-
-    static bool is_valid(const intptr_t *index, const start_stop_t *start_stop) {
-        return (index[0] >= start_stop[0].start) && (index[0] < start_stop[0].stop);
-    }
-
-    static void incr(const char *&pointer, intptr_t *index, const intptr_t *DYND_UNUSED(sizes), const intptr_t *strides) {
-        ++index[0];
-        pointer += strides[0];
-    }
+template <typename T>
+struct is_dim<fixed_dim<T>> {
+  static const bool value = true;
 };
 
-template <typename T, int N>
-class strided_vals {
-    struct {
-        const char *pointer;
-        intptr_t strides[N];
-    } m_data;
-    struct {
-        const char *pointer;
-        intptr_t strides[N];
-    } m_mask;
-    intptr_t m_sizes[N];
-    const start_stop_t *m_start_stop;
-    intptr_t m_center_index[N];
+template <typename T>
+class fixed_dim : public detail::fixed_dim<T, !is_dim<T>::value> {
+  char m_arrmeta[fixed_dim::arrmeta_size];
+  const char *m_data;
+  const start_stop_t *m_start_stop;
```

Comment:
It's a bit vestigal. We need a mechanism to support using fixed_dim in neighborhood. This tells which parts of the neighborhood are out of bounds. I'm trying to figure out a better way to do that...